### PR TITLE
6431 add consumption tracking

### PR DESCRIPTION
**Pull request by j-koehler**
resolves phpstan/phpstan#6431

[As requested](https://github.com/phpstan/phpstan/issues/6431#issuecomment-1023579070) this PR adds consumption tracking and output for both debug runs and standard analyser/parallel analyser (when started with `-vvv`).

(This is a follow up to https://github.com/phpstan/phpstan-src/pull/966 which was closed refactoring / adding worker support.)

### What is added:

* added `make` target for debug mode (increased memory limit already required for runs on `master` branch)
* added helper for converting (fractional) seconds into human readable format
* added per file (memory/time) consumption tracking for both standard and parallel analyser when started with `-vvv`

### How does it look like

image

💡 Top memory consumers are shown only when run in debug mode, because workers/reactphp mess this up in parallel mode.

### Negative impact

In my opinion: not really. Did lots of runs on my local machine (ryzen 3700x on wsl2) for both `master` and this development branch and the  time impact was minimal (+~2seconds on PHPStan compared to `master`), memory impact not even worth mentioning.

I put a lot of effort into this, feedback is appreciated. 🙏

### CODE REVIEW

**Comment by j-koehler** [comment]
So it's 14 days since I opened the PR. Did you find some time to look into this? Could you give me any feedback?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Hi, sorry for keep you waiting, this PR has been in the back of my mind for some time, because I'd welcome the feature, but I'd go differently in implementing it. What I don't like here is the stateful and mutable nature of `ConsumptionTrackingCollector` and `FileConsumptionTracker`. Methods like `getXForLatestFile` that assume that a different method was called on it before are a no-go for me. Merged code needs to look as if I wrote it because it's going to be me who's going to maintain it for years to come.

Also not keeping track of memory in parallel mode is an ommission I'd also like to see solved by this feature.

It seems like it's too much tricky code (like the `purge` methods) to just keep track of a few integers per file.

I don't know what's the next step here - I could describe how I'd like the implementation to look like, but the best description of an implementation is in code - which means I'd have to develop it myself, which means I'd have to give it my priority right now, which it isn't. That's why some PRs in this project and others often find themselves in a limbo state.

Sending a PR often makes the author think "great, I'm helping this project! Can't wait for this to be merged", but in reality it's asking for more time from the maintainer with an uncertain result.

Good first contribution to a new project should usually be in a smaller scope, changing a few lines here and there, not an 800-lines behemoth :)

**Comment by ondrejmirtes** [comment]
Let's continue discussion here since Twitter isn't great for longer texts :)

Reacting to: https://twitter.com/maschinentraum/status/1496396281517948928

I'm a big fan of iteration! But it needs to iterate from "small but still useful" to "great", not from "bad" to "great" :) Generics were implemented over a course of 70+ PRs.

For example - a good first step here would be to discuss and merge how to pass additional data from the workers to the main process. It doesn't need to be read anywhere, it can just be passed in the JSON.

I'd be totally fine with reading and subtracting time and consumed memory before and after this line in `WorkerCommand`:

```php
$fileAnalyserResult = $fileAnalyser->analyseFile($file, $analysedFiles, $registry, null);
```

No additional classes needed!

**Comment by j-koehler** [comment]
Thank you for your feedback. Appreciated!

Yes, it feels a bit overengineered. But I wanted my code to be tested as good as possible and not just pushing arrays around and adding the 20th+21th parameter to some function :wink:  I'm a big fan of DTO instead of arrays. (But in the end still had to be able to have them, to be able to serialise it through the worker/control layer.) 

> What I don't like here is the stateful and mutable nature of `ConsumptionTrackingCollector` and `FileConsumptionTracker`. Methods like `getXForLatestFile` that assume that a different method was called on it before are a no-go for me. 

True, the files could/should be immutable. But I don't really know how you would implement it in a different way. The Collector is this - a collector/aggregator. What do you do, when you don't add things but want informations? Only thing I can think of is returning a null value or something like this. In each use of it, it's pretty much a `ShouldNotHappenException` :wink:  

> Merged code needs to look as if I wrote it because it's going to be me who's going to maintain it for years to come.

I totally understand this and it's probably the best features for PHPStan should be written by yourself or someone who works for years with you.

> Also not keeping track of memory in parallel mode is an ommission I'd also like to see solved by this feature.

And it get's even more complicated. Because tracking the memory you did in your initial implementation in the debug mode just does not work anymore.

> It seems like it's too much tricky code (like the `purge` methods) to just keep track of a few integers per file.

Tbh I don't really understand about the "tricky code" in the `purgeOverflow` method. It's the sorting and slicing you would have to do, no matter what, when you want a sorted Top X list. (The sorting&purging every X files is just nice to have to keep the memory footprint low. I don't think it's really needed anymore but kept it because it does no real harm.)

> I don't know what's the next step here - I could describe how I'd like the implementation to look like, but the best description of an implementation is in code - which means I'd have to develop it myself, which means I'd have to give it my priority right now, which it isn't. That's why some PRs in this project and others often find themselves in a limbo state.
> 
> Sending a PR often makes the author think "great, I'm helping this project! Can't wait for this to be merged", but in reality it's asking for more time from the maintainer with an uncertain result.

This is totally fine. Some feedback helps not feeling ghosted. Even when it's "I have to think about this." or "busy right now, aiming for next minor update" or even "I don't like this and that, but can't really explain why". You are the maintainer :slightly_smiling_face: 

> Good first contribution to a new project should usually be in a smaller scope, changing a few lines here and there, not an 800-lines behemoth :)

Before creating the first PR I did this exactly: I added the processing time to the debug output. This solution brought me instant benefit, solved my need (-> the issue I created first) and helped me identifying problematic files in my source code. But you requested way more for then this for a first implementation and I thought: well, that's surprising, but I like the idea behind this - challenge accepted and I took an even deeper dive (e.g. how does the parallel mode work).

So this (at least to me, no offence) where you are (kinda) acting in two different ways: You say "start with something small and simple" but ask for "make it big and fully implemented". 
You request advanced or very advanced features (top consumers by type and working in parallel mode and working memory tracking working in parallel mode) while the existing super simple implementation (e.g. current memory tracking in debug mode only) is fine but should not enhanced anymore (e.g. with time tracking). Thus - raising the bar to some degree where probably nobody but yourself can solve it in a way you like.

Don't get me wrong, I understand the this big vision of this feature and it is really nice. I just think iterations and small improvements help as well. (I guess you fear to be left alone with it after some time.)

But considering 

> Merged code needs to look as if I wrote it because it's going to be me who's going to maintain it for years to come.

and

> I could describe how I'd like the implementation to look like, but the best description of an implementation is in code - which means I'd have to develop it myself

I think it's the best to just close this PR.

I have a fork working for my very own needs. It does not help anybody else, but I understand this is not good enough for the "main" PHPStan. But to me the task looks impossible without wasting each others time (me trying to implement your vision in a way you would implement it and you reviewing, discussing and not liking it). 

I'll playing around with your feedback (immutable) and with the memory tracking in parallel mode out of curiosity but that's about it.

**Comment by j-koehler** [comment]
Woops, you added something while I wrote my lengthy feedback comment. Reopen.

**Comment by j-koehler** [comment]
> I'd be totally fine with reading and subtracting time and consumed memory before and after this line in WorkerCommand:

This works for time tracking yes. But it looks to me that this is not possible for memory - probably because of how ReactPHP works. I tried using the simple logic, but numbers became garbage. So this is why I removed it for now (Again: this feels to me like this could be an iteration: Add memory tracking for parallel mode)

> No additional classes needed!

True. I added the classes mainly to be able to test&verify my own added functionality. When just pushing arrays around this is not testable to me.

**Pull request closed without merging**
